### PR TITLE
Changes to analyzer and result producer

### PR DESCRIPTION
**Pull request by gmadigan**

### CODE REVIEW

File: LQResultProducer.py [hunk 1]
```diff
@@ -40,15 +41,16 @@
 TreeName = "PhysicalVariables"
 
 # Integrated luminosity for normalization
+# Udated July 13th 2021: https://twiki.cern.ch/twiki/bin/viewauth/CMS/TWikiLUM
 if year == '2016':
-	lumi = 35920.
-	lumiInvfb = '35.9'
+	lumi = 36330. #35920.
+	lumiInvfb = '36.33' #'35.9'
```
Comment (DavidMorse) [inline LQResultProducer.py]: Please leave this as 1 decimal place, 36.3, 41.5, 59.8

File: LQResultProducer.py [hunk 2]
```diff
@@ -140,8 +144,8 @@
 #NormalWeightMuMu = str(lumi)+'*weight_central*((pass_HLTMu50+pass_HLTTkMu50)>0)'+doublemuHLT+doubleMuIdScale+doubleMuIsoScale+trackerHIP1+trackerHIP2
 #NormalWeightMuNu = str(lumi)+'*weight_central*((pass_HLTMu50+pass_HLTTkMu50)>0)'+singlemuHLT+singleMuIdScale+singleMuIsoScale+trackerHIP1
 #NormalWeightEMu = str(lumi)+'*weight_central*((pass_HLTMu50+pass_HLTTkMu50)>0)'+singlemuHLTEMU+MuIdScaleEMU+MuIsoScaleEMU+eleRECOScale+eleHEEPScale+trackerHIPEMU
-NormalWeightMuMu = str(lumi)+'*weight_central'+doublemuHLT+doubleMuIdScale+doubleMuIsoScale+trackerHIP1+trackerHIP2+bTagSFmedium
-NormalWeightMuNu = str(lumi)+'*weight_central'+singlemuHLT+singleMuIdScale+singleMuIsoScale+trackerHIP1
+NormalWeightMuMu = str(lumi)+'*weight_central'+doublemuHLT+doubleMuRecoSF+trackerHIP1+trackerHIP2+bTagSFmedium
```
Comment (DavidMorse) [inline LQResultProducer.py]: why are ID and ISO SF not here?

File: LQResultProducer.py [hunk 3]
```diff
@@ -140,8 +144,8 @@
...
+NormalWeightMuNu = str(lumi)+'*weight_central'+singlemuHLT+singleMuRecoSF+trackerHIP1
```
Comment (DavidMorse) [inline LQResultProducer.py]: why are ID and ISO SF not here?

File: LQResultProducer.py [hunk 4]
```diff
@@ -342,11 +346,11 @@ def main():
 	# for this, and make use of it. e.g. For systematic variations, we can run in batch instead
 	# of running serially, which speeds things up.
 
-	version_name = 'Testing_'+year+'_stockNano' # scriptflag
+	version_name = 'Testing_'+year+'_stockNanoAODv7_HighPtIso' # scriptflag
 	#version_name = 'Testing_noQCD_14nov' # use sf tag above if this is the real folder
 	os.system('mkdir Results_'+version_name) 
 
-	MuMuOptCutFile = 'Results_'+version_name+'/OptLQ_uujjCuts_Smoothed_pol2cutoff.txt' # scriptflag
+	MuMuOptCutFile = 'Results_'+version_name+'/OptLQ_uujjCuts.txt' # scriptflag
```
Comment (DavidMorse) [inline LQResultProducer.py]: please leave the _Smoothed_pol2cutoff naming convention, it is useful later

File: LQResultProducer.py [hunk 5]
```diff
@@ -981,7 +1029,8 @@ def main():
 		scaleFactors = [Rz_uujj,Rtt_uujj,Rw_uvjj]
 		if not doLongLived :
 			MuMuOptTestCutFile = 'Results_'+version_name+'/OptLQ_uujjCuts_Smoothed_pol2cutoff.txt'
-			variableSpace = ['M_uu:25:100:1000','St_uujj:100:300:2500','M_uujj2:25:100:1000']
+			#variableSpace = ['M_uu:25:100:1000','St_uujj:100:300:2500','M_uujj2:25:100:1000']
+			variableSpace = ['M_uujj:25:500:2500','St_uujj:100:300:2500',dR_muon1muon2_jet1+':0.5:0:10']
```
Comment (DavidMorse) [inline LQResultProducer.py]: I guess this was an experiment? Please remove it and uncomment the previous line

File: LQResultProducer.py [hunk 6]
```diff
@@ -3963,10 +4012,18 @@ def ModSelection(selection,sysmethod,channel_log):
...
 		#Some updates:
 		#https://indico.cern.ch/event/675475/contributions/2764498/subcontributions/240732/attachments/1547347/2429001/Wprime-muon-Approvalv1.pdf
+
+		#Percent systematic uncertainties of integrated luminosity provided in table, here: https://twiki.cern.ch/twiki/bin/viewauth/CMS/TWikiLUM
 		if sysmethod == 'LUMIup':
-			selection = '(1.025)*'+selection
+			if year =='2016': selection = '(1.012)*'+selection #vary 2016 luminosity up by 1.2%
+			elif year == '2017': selection = '(1.023)*'+selection #vary 2017 luminosity up by 2.3%
+			elif year == '2018': selection = '(1.025)*'+selection #vary 2018 luminosity up by 2.5%
+			#selection = '(1.025)*'+selection
```
Comment (DavidMorse) [inline LQResultProducer.py]: please delete this line

File: LQResultProducer.py [hunk 7]
```diff
@@ -3963,10 +4012,18 @@ def ModSelection(selection,sysmethod,channel_log):
...
 		if sysmethod == 'LUMIdown':
-			selection = '(0.975)*'+selection
+			if year =='2016': selection = '(0.988)*'+selection #vary 2016 luminosity down by 1.2%
+			elif year == '2017': selection = '(0.977)*'+selection #vary 2017 luminosity down by 2.3%
+			elif year == '2018': selection = '(0.975)*'+selection #vary 2018 luminosity down by 2.5%
+			#selection = '(0.975)*'+selection
```
Comment (DavidMorse) [inline LQResultProducer.py]: please delete this line

File: LQResultProducer.py [hunk 8]
```diff
@@ -4042,8 +4099,28 @@ def SysTable(optimlog, selection_uujj,selection_uvjj,NormalDirectory, weight,sys
...
+	#[[Rz_uujj,Rz_uujj_err],[Rtt_uujj,Rtt_uujj_err]] = GetMuMuScaleFactors( selection_uujj, NormalDirectory, '(M_uu>80)*(M_uu<100)', '(M_uu>100)*(M_uu<250)',0,0)
+	#if year == '2016':
+	#	[[Rz_uujj,Rz_uujj_err],[Rtt_uujj,Rtt_uujj_err]] = [[1.019,0.017],[1.134,0.013]] #2016 stock NanoAODv7 with 1 btag
+	#else:
+	#	[[Rz_uujj,Rz_uujj_err],[Rtt_uujj,Rtt_uujj_err]] = GetMuMuScaleFactors( NormalWeightMuMu+'*'+preselectionmumu, NormalDirectory, '(M_uu>80)*(M_uu<100)', '(M_uu>100)*(M_uu<250)',0,0)
+	#[[Rw_uvjj,Rw_uvjj_err],[Rtt_uvjj,Rtt_uvjj_err]] = [[1.0,0.0],[1.0,0.0]]
+	#[[Rw_uvjj,Rw_uvjj_err],[Rtt_uvjj,Rtt_uvjj_err]] = GetMuNuScaleFactors( selection_uvjj, NormalDirectory, munu1,munu2,0)
+
+	# SFs with updated integrated luminosities
+	if year == '2016':
```
Comment (DavidMorse) [inline LQResultProducer.py]: you can't hardcode these.....they will change with the systematic variations. You have to let them recalculate. But that also takes out the yearly differences, so the code can be simpler, I think just reverted to the original, with the correct TTBar normalization control region (100-250)

File: LQResultProducer.py [hunk 9]
```diff
@@ -7922,7 +8010,7 @@ def FixFinalCards(cardsets):
 	for c in cardsets:
 		for line in open(c,'r'):
 			line = line.replace('uujj','_M_')
-			line = line.replace('uvjj','_BetaHalf_M_')
+			#line = line.replace('uvjj','_BetaHalf_M_')
```
Comment (DavidMorse) [inline LQResultProducer.py]: no need to comment this out, please remove the #

File: NTupleAnalyzer_nanoAOD.py [hunk 10]
```diff
@@ -2290,8 +2311,8 @@ def GetLLJJMasses3Jets(l1,l2,j1,j2,j3):
 	t.GetEntry(n)
 	# if n > 1000:  # Testing....
 	# 	break
-	if n%100==0:
-		print 'Processing event',n, 'of', N # where we are in the loop...
+	#if n%100==0:
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: why turn this off? if output is too large it could e.g. be changed to 1000 instead of 100

**Review by DavidMorse — CHANGES_REQUESTED** [review]

File: LQResultProducer.py (hunk 6 above)
Comment (gmadigan) [reply LQResultProducer.py]: Which line(s) exactly? 4021?

File: LQResultProducer.py (hunk 7 above)
Comment (gmadigan) [reply LQResultProducer.py]: Which line(s) exactly? 4026?

File: LQResultProducer.py (hunk 2 above)
Comment (gmadigan) [reply LQResultProducer.py]: I thought (after a re-read, and seeing no SF table) the recommendation was that they need not be applied, but I think the phrasing is a little confusing (e.g., "No additional treatment").
https://twiki.cern.ch/twiki/bin/view/CMS/HighPtMuonReferenceRun2#Isolation
https://twiki.cern.ch/twiki/bin/view/CMS/HighPtMuonReferenceRun2#Identification
Also that would contradict: https://twiki.cern.ch/twiki/bin/view/CMS/MuonLegacy2016#ID_and_ISO_efficiencies
I'll add them back in.

Comment (DavidMorse) [reply LQResultProducer.py]: Hmm ok......let's follow up, a message to the Muon POG hypernews can clear it up quickly, just say what ID and ISO we use, point to the Twiki and say it looks like we should not apply SFs, is this correct?

File: LQResultProducer.py (hunk 6 above)
Comment (DavidMorse) [reply LQResultProducer.py]: yes 4021, the out of date one

File: LQResultProducer.py (hunk 7 above)
Comment (DavidMorse) [reply LQResultProducer.py]: yes

File: NTupleAnalyzer_nanoAOD.py [hunk 11]
```diff
@@ -1114,31 +1117,118 @@ def getMuonSF(_pt,_eta):
...
+			origLorentz = TLorentzVector()
+			origLorentz.SetPtEtaPhiM(_ptCollection[n], _etaCollection[n], _phiCollection[n], 0)
+			#print 'origLorentz is',origLorentz
+			# Smearing is performed on 3-momentum, convert from Pt, Eta, Phi to cartesian 4-momentum and back
+			origPx = origLorentz.Px()
+			origPy = origLorentz.Py()
+			origPz = origLorentz.Pz()
+
+			# Smear momenta here
+			smearedPx = origPx*(1 + tRand.Gaus(0.0, MERParametrization(origPx, _etaCollection[n])*smearConst))
```
Comment (DavidMorse) [inline NTupleAnalyzer_nanoAOD.py]: So this smears px,py,pz by different amounts. Shouldn't they all be smeared by the same amount? Did you get this code example from somewhere?

**Review by DavidMorse — CHANGES_REQUESTED** [review]

File: NTupleAnalyzer_nanoAOD.py (hunk 11 above)
Comment (gmadigan) [reply NTupleAnalyzer_nanoAOD.py]: I had to code this myself. You're right though, and it should be changed to smear each p component by the same amount. I'll fix that

**Review by DavidMorse — APPROVED** [review]

**Pull request merged**